### PR TITLE
Refactor TransportShardReplicationOperationAction

### PR DESCRIPTION
**Pull request by bleskes**
Refactors TransportShardReplicationOperationAction state management into clear separate Primary phase and Replication phase. The primary phase is responsible for routing the request to the node  holding the primary, validating it and performing the operation on the primary. The Replication phase is responsible for sending the request to the replicas and managing their responses.

This refactoring is aimed at simplifying adding operation start and end hooks that are needed for the counter mentioned in #10032 . 

This also adds unit test infrastructure for this class, and some basic tests. We can extend later as we continue developing.

For now, this is planned to go to 2.0.0 but we make backport it to 1.6.0, depending how work goes with #10032

### CODE REVIEW

**Comment by bleskes** [comment]
@martijnvg @brwe can you take a look?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 1]
```diff
@@ -302,149 +301,159 @@ public void writeTo(StreamOutput out) throws IOException {
...
+                    @Override
+                    public void handleException(TransportException exp) {
+                        // if we got disconnected from the node, or the node / shard is not in the right state (being closed)
+                        if (exp.unwrapCause() instanceof ConnectTransportException || exp.unwrapCause() instanceof NodeClosedException ||
+                                retryPrimaryException(exp)) {
+                            internalRequest.request().setCanHaveDuplicates();
+                            // we already marked it as started when we executed it (removed the listener) so pass false
+                            // to re-add to the cluster listener
+                            logger.trace("received an error from node the primary was assigned to ({}), scheduling a retry", exp.getMessage());
+                            retry(exp);
```
Comment (brwe) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: what if retry() throws an exception?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 2]
```diff
@@ -631,34 +791,34 @@ void performOnReplica(final ReplicationState state, final ShardRouting shard, fi
...
+                                    logger.warn("failed to perform " + actionName + " on remote replica " + node + shardIt.shardId(), exp);
+                                    shardStateAction.shardFailed(shard, indexMetaData.getUUID(),
+                                            "Failed to perform [" + actionName + "] on replica, message [" + ExceptionsHelper.detailedMessage(exp) + "]");
+                                }
+                            }
 
-                });
+                        });
             } else {
-                if (internalRequest.request().operationThreaded()) {
```
Comment (brwe) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: can we remove this whole block and add a check before that the node id is not the local one while we are already at it or do it in a different pr?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 3]
```diff
@@ -519,101 +661,119 @@ void performReplicas(PrimaryOperationRequest por, Tuple<Response, ReplicaRequest
             ClusterState newState = clusterService.state();
             ShardRouting newPrimaryShard = null;
```
Comment (brwe) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: not used?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 4]
```diff
@@ -461,7 +470,7 @@ void retry(Throwable failure) {
             observer.waitForNextChange(new ClusterStateObserver.Listener() {
                 @Override
                 public void onNewClusterState(ClusterState state) {
-                    doStart();
+                    start();
                 }
 
                 @Override
```
Comment (brwe) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: we should call finishAsFailed() also in onClusterServiceClose() I think?

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+        assertFalse("primary phase should stop execution on retryable block", primaryPhase.checkBlocks());
+        assertFalse("primary phase should wait on retryable block", listener.isDone());
+
+        block = ClusterBlocks.builder()
+                .addGlobalBlock(new ClusterBlock(1, "non retryable", false, true, RestStatus.SERVICE_UNAVAILABLE, ClusterBlockLevel.ALL));
+        clusterService.setState(ClusterState.builder(clusterService.state()).blocks(block));
+        assertListenerThrows("primary phase should fail operation when moving from a retryable block a non-retryable one", listener, ClusterBlockException.class);
+    }
+
+    ClusterState stateWithUnassingedShards(String index, int numberOfReplicas) {
```
Comment (brwe) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: not used?

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+        ShardRoutingState[] replicaStates = new ShardRoutingState[numberOfReplicas];
+        for (int i = 0; i < replicaStates.length; i++) {
+            replicaStates[i] = ShardRoutingState.UNASSIGNED;
+        }
+        return state(index, false, ShardRoutingState.UNASSIGNED, replicaStates);
+    }
+
+
+    ClusterState stateWithStartedPrimary(String index, boolean primaryLocal, int numberOfReplicas) {
+        ShardRoutingState primaryState;
```
Comment (brwe) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: not used?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 7]
```diff
@@ -472,43 +481,176 @@ public void onClusterServiceClose() {
...
+        void finishAndMoveToReplication(ReplicationPhase replicationPhase) {
+            if (finished.compareAndSet(false, true)) {
+                replicationPhase.start();
+            } else {
+                assert false : "finishAndMoveToReplication called but operation is already finished";
+            }
+        }
+
+
+        void finishAsFailed(Throwable failure) {
```
Comment (brwe) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: Is this supposed to be the hook for the counter dec? It seems to be called after remote execution failed (https://github.com/elastic/elasticsearch/pull/10749/files#diff-a1076062ccd18e89873d62fe1d78e1e1R453) sometimes when the operation could not be executed at all due to blocks (https://github.com/elastic/elasticsearch/pull/10749/files#diff-a1076062ccd18e89873d62fe1d78e1e1R363) and also if write locally on the primary failed (https://github.com/elastic/elasticsearch/pull/10749/files#diff-a1076062ccd18e89873d62fe1d78e1e1R408) I would still have to be as careful with where to increment and decrement the counter like before.

I was hoping that this refactoring separates remote and local execution for the primary operation more strictly because this is I think where the major difficulty with reasoning about what the code does stems from. I might not fully understand though...

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 8]
```diff
@@ -472,43 +481,176 @@ public void onClusterServiceClose() {
...
+        private final ShardRouting originalPrimaryShard;
+        private final AtomicInteger pending;
+        private final int totalShards;
+        private final ClusterStateObserver observer;
+
+        /**
+         * the constructor doesn't take any action, just calculates state. Call {@link #start()} to start
+         * replicating.
+         */
+        public ReplicationPhase(ShardIterator originalShardIt, ReplicaRequest replicaRequest, Response finalResponse,
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: maybe make this package protected?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 9]
```diff
@@ -519,101 +661,119 @@ void performReplicas(PrimaryOperationRequest por, Tuple<Response, ReplicaRequest
...
-                // (and we already did it on the primary)
-                boolean doOnlyOnRelocating = false;
-                if (shard.primary()) {
-                    if (shard.relocating()) {
-                        doOnlyOnRelocating = true;
+                    // we index on a replica that is initializing as well since we might not have got the event
+                    // yet that it was started. We will get an exception IllegalShardState exception if its not started
+                    // and that's fine, we will ignore it
+                    if (shard.primary()) {
+                        if (originalPrimaryShard.currentNodeId().equals(shard.currentNodeId()) == false) {
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: if this return `true` then that would be odd? maybe we should log this and add an assert?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 10]
```diff
@@ -472,43 +481,176 @@ public void onClusterServiceClose() {
...
+    }
+
+    /** inner class is responsible for send the requests to all replica shards and manage the responses */
+    public final class ReplicationPhase {
+
+        private final ReplicaRequest replicaRequest;
+        private final Response finalResponse;
+        private final ShardIterator shardIt;
+        private final ActionListener<Response> listener;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I find it confusing the we have the same field names for this in both ReplicationPhase and PrimaryPhase.

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 11]
```diff
@@ -623,7 +783,7 @@ void performOnReplica(final ReplicationState state, final ShardRouting shard, fi
                 // to wait until they get the new mapping through the cluster
                 // state, which is why we recommend pre-defined mappings for
                 // indices using shadow replicas
-                state.onReplicaSuccess();
+                onReplicaSuccess();
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: just an idea: we could skip the creation of the `ReplicationPhase` entirely if shadow replicas are being used.

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.cluster.metadata.IndexMetaData.*;
+import static org.hamcrest.Matchers.*;
+
+public class ShardReplicationOperationTests extends ElasticsearchTestCase {
```
Comment (martijnvg) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: big +1 for this unit test!

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+    private static ThreadPool threadPool;
+
+    private TransportService transportService;
```
Comment (martijnvg) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: this can become a local variable in setUp()

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+        assertThat(shardInfo.getFailures(), arrayWithSize(criticalFailures));
+        assertThat(shardInfo.getSuccessful(), equalTo(successfull));
+        assertThat(shardInfo.getTotal(), equalTo(shardRoutingTable.size()));
+
+        assertThat("failed to see enough shard failures", transport.capturedRequests().length, equalTo(criticalFailures));
+        for (CapturingTransport.CapturedRequest capturedRequest : transport.capturedRequests()) {
+            assertThat(capturedRequest.action, equalTo(ShardStateAction.SHARD_FAILED_ACTION_NAME));
+        }
+    }
+
```
Comment (martijnvg) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: maybe we add a test that tests what happens if not enough shard copies are available (write consistency)?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 10 above)
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: actually I just got a bit confused because both classes are in the same file...

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 15]
```diff
@@ -302,149 +301,159 @@ public void writeTo(StreamOutput out) throws IOException {
...
+        PrimaryPhase(Request request, ActionListener<Response> listener) {
             this.internalRequest = new InternalRequest(request);
             this.listener = listener;
+            this.observer = new ClusterStateObserver(clusterService, internalRequest.request().timeout(), logger);
         }
 
         public void start() {
-            this.observer = new ClusterStateObserver(clusterService, internalRequest.request().timeout(), logger);
-            doStart();
+            try {
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: make start() package protected?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 16]
```diff
@@ -472,43 +481,176 @@ public void onClusterServiceClose() {
...
+            if (indexShard == null) {
+                logger.debug("ignoring failed replica [{}][{}] because index was already removed.", index, shardId);
+                return;
+            }
+            indexShard.failShard(actionName + " failed on replica", t);
         }
+    }
+
+    /** inner class is responsible for send the requests to all replica shards and manage the responses */
+    public final class ReplicationPhase {
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: package protected?

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 11 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: ++ . it should be ignored in the counting phase..

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 2 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I rather do it in another change.

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 7 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: my main concern was the lack of separation between the part that sends remote replication request and the rest of the code. This is what I separated here. In the future we will remove all local execution and use the short cut in the transport service

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 8 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: ++

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 9 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: it means the shard relocated beneath our feet. It's very rare and I don't think we do the right thing here per se, but that is what we were doing and I rather not change it for now...

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 11 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I change shadow replicas handling. I think it's easier to still keep them here, for the case of a relocating primary for example

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java (hunk 13 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: changed

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java (hunk 14 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: added

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 15 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I changed it to be AbstractRunnable for other reasons.. the class is package private now

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 16 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: done

**Comment by bleskes** [comment]
@martijnvg @brwe pushed an update with all the feedback + extra testing.

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 17]
```diff
@@ -583,82 +781,69 @@ void performReplicas(PrimaryOperationRequest por, Tuple<Response, ReplicaRequest
...
-                            logger.warn("Failed to perform " + actionName + " on remote replica " + node + shardIt.shardId(), exp);
-                            shardStateAction.shardFailed(shard, indexMetaData.getUUID(),
-                                    "Failed to perform [" + actionName + "] on replica, message [" + ExceptionsHelper.detailedMessage(exp) + "]");
-                        }
-                    }
+                            @Override
+                            public void handleException(TransportException exp) {
+                                onReplicaFailure(nodeId, exp);
+                                logger.trace("[{}] transport failure during replica request [{}] ", exp, node, replicaRequest);
+                                if (!ignoreReplicaException(exp)) {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: can we use `== false`

File: src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java (hunk 12 above)
Comment (s1monw) [inline src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java]: yeah it's really good!

**Comment by s1monw** [comment]
I have to review the logic agian but the test alone makes me +1 this

**Comment by martijnvg** [comment]
LGTM

**Comment by bleskes** [comment]
pushed to 1.x as well.
